### PR TITLE
67 - audit history feature for accounts

### LINKED ISSUE #67: Add audit / history feature for accounts
Users must be able to see a list of changes that have occurred to their account. The following field should be captured.

1. What happened.
2. When did it happen.
3. Who did it.

### PR DESCRIPTION
**Pull request by piotrkosecki**
resolves #67

### CODE REVIEW

**Review by ivanopagano — DISMISSED** [review]
I have only minor comments on possible improvements, I left them for your evaluation.

The logic looks good to me.

File: doc/nautilus.sql [hunk 1]
```diff
@@ -78,6 +78,14 @@ create TABLE metering_statistics(
 
 CREATE UNIQUE INDEX metering_statistics_idx ON metering_statistics(userid, period_start, period_end);
 
+create TABLE user_history(
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply doc/nautilus.sql:81]: For me it's ok, but lets ask @bhaskarkishore

File: src/main/scala/tech/cryptonomic/nautilus/cloud/adapters/akka/UserActionHistoryOperations.scala [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        Some(session.userId)
+      } else
+        Try {
+          path.stripPrefix("/users/").takeWhile(_ != '/').toInt
+        }.toOption
+
+      val address = ip.toIP.map(_.ip.getHostAddress)
+      val userAction = maybeUserId.map(u => UserAction(u, Some(session.userId), Instant.now(), address, path))
+
+      request.method match {
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/nautilus/cloud/adapters/akka/UserActionHistoryOperations.scala:21]: sure, changed

File: src/main/scala/tech/cryptonomic/nautilus/cloud/adapters/akka/UserActionHistoryOperations.scala [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import tech.cryptonomic.nautilus.cloud.domain.user.history.{UserAction, UserHistoryService}
+
+import scala.util.Try
+
+/** Class with methods for gathering user history */
+class UserActionHistoryOperations(userHistoryService: UserHistoryService[IO]) {
+
+  /** Directive for logging user actions to thee DB for non-OPTIONS/GET requests */
+  def logRequestWithSession(ip: RemoteAddress)(implicit session: Session): Directive[Unit] =
+    BasicDirectives.extractRequest.flatMap { request =>
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/nautilus/cloud/adapters/akka/UserActionHistoryOperations.scala]: yes, you're right, I was too caught up in the logic that I've missed obvious improvement, good catch!

File: src/main/scala/tech/cryptonomic/nautilus/cloud/adapters/akka/UserActionHistoryOperations.scala [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+      request.method match {
+        case HttpMethods.OPTIONS | HttpMethods.GET => ()
+        case _ => userAction.foreach(ua => userHistoryService.insertUserAction(ua).unsafeRunAsyncAndForget())
+      }
+      BasicDirectives.mapResponse(identity)
+    }
+
+  /** Generic log for user action */
+  def logReqest(userId: UserId, ip: Option[String], action: String): Unit = {
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/nautilus/cloud/adapters/akka/UserActionHistoryOperations.scala]: good idea, changed

**Review by ivanopagano — APPROVED** [review]
For me it's basically good.

The main leftover comment is about table naming, yet @bhaskarkishore will choose, and then merge by himself, so I claim it done for me.

**Review by bhaskarkishore — APPROVED** [review]

**Comment by bhaskarkishore** [comment]
Closing for now.

**Pull request closed without merging**